sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -210,6 +243,10 @@ class CompoundBodyExec(statements: Seq[CompoundStatementExec], label: Option[Str
             statement
           case Some(body: NonLeafStatementExec) =>
             if (body.getTreeIterator.hasNext) {
+              body match {
+                case exec: CompoundBodyExec => exec.enterScope()
```

Comment:
nit: maybe a small comment explaining that this will actually be executed only once per scope, due to the logic in `enterScope`